### ISSUE TITLE
regression in list assignment with undef in list context

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16685]
Migrated from [rt.perl.org#133503](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133503) (status was 'open')

Searchable as RT133503$

**Comment by p5pRT** [comment]
### From wolf-dietrich_moeller@t-online.de

The documentation on assignments in perlop reads:
"a list assignment in list context produces the
list of lvalues assigned to"

The following program shows a regression,
when the "list assigned to" contains "undef".
(tested with two old versions available to me:
- Strawberry Perl Portable 5.22.3.1-32bit
- Strawberry Perl Portable 5.26.2.1-32bit)

The newer version 5.26 (and also 5.28) works correctly if the
third element in the list assigned to is a variable (test 3),
but ignores the third element if it is undef.
Thus `@s` in test 4 only has 2 elements instead of 3.

```perl
my ($y, @s);
my $x = 5;
print 'test 1: ',scalar (@s = (($y) = (8,9))),' ',$y,' ',$x//'undef',"\n";
my $x = 5;
print 'test 2: ',scalar (@s = (($y,undef) = (8,9))),' ',$y,'
',$x//'undef',"\n";
my $x = 5;
print 'test 3: ',scalar (@s = (($y,undef,$x) = (8,9))),' ',$y,'
',$x//'undef',"\n";
my $x = 5;
print 'test 4: ',scalar (@s = (($y,undef,undef) = (8,9))),' ',$y,'
',$x//'undef',"\n";
```

```
Result in 5.22 (correct, @s == 3):
test 1: 1 8 5
test 2: 2 8 5
test 3: 3 8 undef
test 4: 3 8 5
```

```
Result in 5.26 (test 4 incorrect, @s == 2):
test 1: 1 8 5
test 2: 2 8 5
test 3: 3 8 undef
test 4: 2 8 5
```

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Sep 07\, 2018 at 08&#8203;:48&#8203;:29AM \-0700\, Wolf\-Dietrich Moeller \(via RT\) wrote&#8203;:
> The documentation on assignments in perlop reads&#8203;:
> "a list assignment in list context produces the
>  list of lvalues assigned to"
> 
> The following program shows a regression\,
> when the "list assigned to" contains "undef"\.
> \(tested with two old versions available to me&#8203;:
> \- Strawberry Perl Portable 5\.22\.3\.1\-32bit
> \- Strawberry Perl Portable 5\.26\.2\.1\-32bit\)
> 
> The newer version 5\.26 \(and also 5\.28\) works correctly if the
> third element in the list assigned to is a variable \(test 3\)\,
> but ignores the third element if it is undef\.
> Thus @&#8203;s in test 4 only has 2 elements instead of 3\.

Bisects to this\. I haven't looked further yet\.

commit b09ed995add057b8e0b51964b48ef1d1cc3c9c91
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Wed Oct 19 09&#8203;:41&#8203;:53 2016 \+0100
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Wed Oct 26 08&#8203;:37&#8203;:27 2016 \+0100

&nbsp;    Handle list assignment in list context better
&nbsp;    
&nbsp;    In something like
&nbsp;    
&nbsp;        sub inc \{ $\_\+\+ for @&#8203;\_ \}
&nbsp;        inc\(\($a\,$b\,$c\,$d\) = \(10\,20\)\)
&nbsp;    
&nbsp;    The four scalar vars will end up with the values \(11\,21\,1\,1\)\, with
&nbsp;    the list assign op returning 4 lval scalars to be passed to inc\.
&nbsp;    
&nbsp;    However\, when the LHS includes a hash or array\, any 'empty' scalars weren't
&nbsp;    being returned\. This&#8203;:
&nbsp;    
&nbsp;        inc\(\($a\,$b\,@&#8203;array\,$c\) = \(10\)\)
&nbsp;    
&nbsp;    used to leave $b and $c undefined\. With this commit\, they are both set to
&nbsp;    1\.
&nbsp;    
&nbsp;    This change broke some tests in hashassign\.t\, which were added in 2012
&nbsp;    by commit v5\.17\.6\-295\-gb1babc5\. Half these tests were commented as
&nbsp;    
&nbsp;        \# this arguable\, but this is how it works
&nbsp;    
&nbsp;    and they all tested that a scalar following a hash wasn't returned in
&nbsp;    lvalue context; i\.e\. they all assumed that
&nbsp;    
&nbsp;        inc\(\(%h\, $x\) = \(\.\.\.\)\)
&nbsp;    
&nbsp;    \*wouldn't\* increment $x\. This commit causes $x to be incremented\, and
&nbsp;    I've changed the failing tests\.
&nbsp;    
&nbsp;    It also adds an EXTEND\(SP\,1\) in the scalar case; otherwise with
&nbsp;    scalar\( \(\) = @&#8203;a\) and empty @&#8203;a\, it could in theory push the '0' return
&nbsp;    value off the end of the stack\.


\-\- 
Spock \(or Data\) is fired from his high\-ranking position for not being able
to understand the most basic nuances of about one in three sentences that
anyone says to him\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#19

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by toddr** [comment]
This is unchanged in 5.30

**Comment by iabyn** [comment]
Now fixed in blead by commit v5.33.0-236-g282d9dfeb4
